autoprecompiles/src/optimizer_documentation.md

```diff
@@ -204,6 +204,63 @@ TODO Continue with the abstraction using Range Constraints.
 ## Combining Range Constraints
 
 
+## Grouped Expressions
+
+The main data structure used for algebraic expressions the _Grouped Expression_.
+A Grouped Expression consists of a constant term, a list of linear terms (a list of pairs of a non-zero coefficient
+and a variable) and a list of quadratic terms (a list of pairs of Grouped Expressions).
+
+For affine Algebraic Expressions, the Grouped Expression provides a normal form (if sorted by variable).
+It even provides a normal form for affine Algebraic Constraints if multiplied by a factor such that
+the constant term is zero or one.
+
+This normal fom makes it easy to compare affine Algebraic Expressions for equality,
+check if they have a constant difference, etc.
+
+If a map data structure is used for the linear terms, substitution of variables is also quick.
+
+Addition and subtraction of Grouped Expressions is implemented to remove linear terms that cancel each other out,
+and it performs some checks also in the quadratic terms, but this part is not complete for performance reasons.
+
+## Equivalence Notion
+
+TODO define the property of a Constraint System the optimizer preserves
+
+We call two Constraint Systems _equivalent_ if every satisfying assignment for one system
+can be extended to a satisfying assignment for the other system and every such extension
+leads to the same payloads and multiplicities for all _stateful_ bus interactions in both systems.
```

Comment:
And every non-satisfying assignment of the first system must also not satisfy the second